### PR TITLE
Add __env__ substitution inside file and pillar root paths

### PR DESCRIPTION
**Pull request by nicholasmhughes**
### What does this PR do?
Adds `__env__` substitution inside file and pillar root paths

### What issues does this PR fix or reference?
Addresses the spirit of this issue: #55747, but not the exact implementation of the resulting PR (#55812)... which may be abandoned at this time.

### Previous Behavior
The `__env__` designation could be used as a key underneath the `file_roots` or `pillar_roots` configuration to provide a "catch-all" root location to provide minions with random `saltenv` and/or `pillarenv` set. The configuration below would send minions with `saltenv` set to `base`, `foo`, and `bar` to the same location:

```yaml
file_roots:
  __env__:
    - /srv/salt
```

However, there wasn't a way to dynamically set a different path location for each environment. The configuration below would be literally interpreted as `/srv/__env__/salt` for each environment.

```yaml
file_roots:
  __env__:
    - /srv/__env__/salt
```

### New Behavior
This PR takes dynamic environments one step further so `__env__` can also be used in the `file_roots` and `pillar_roots` paths. It will be replaced with the actual `saltenv` or `pillarenv` and searched for states and/or data to provide to the minion. For instance, this configuration:

```yaml
file_roots:
  __env__:
    - /srv/__env__/salt
```

is now equivalent to this static configuration:

```yaml
file_roots:
  base:
    - /srv/base/salt
  foo:
    - /srv/foo/salt
  bar:
    - /srv/bar/salt
```

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
Yes

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

### CODE REVIEW

File: salt/fileserver/roots.py [hunk 1]
```diff
@@ -94,6 +95,7 @@ def _add_file_stat(fnd):
             return _add_file_stat(fnd)
         return fnd
     for root in __opts__["file_roots"][saltenv]:
+        root = root.replace("__env__", actual_saltenv)
```
Comment (MKLeb) [inline salt/fileserver/roots.py:98]: Would it make more sense to use something like this?
```suggestion
        if root == "__env__":
            root = actual_saltenv
```
That way it won't replace `__env__` in other environments where users may be using that as a prefix/something similar.

Comment (nicholasmhughes) [reply salt/fileserver/roots.py:98]: I kinda like the flexibility of being able to replace for any env, but I definitely get your point. I don't want to break anyone. I'll see about that change.

Comment (nicholasmhughes) [reply salt/fileserver/roots.py]: @MKLeb , I changed the logic to only substitute `__env__` found in the root paths of the `__env__` environment.

**Comment by nicholasmhughes** [comment]
re-run all

**Comment by nicholasmhughes** [comment]
re-run all

**Comment by nicholasmhughes** [comment]
re-run all

**Comment by nicholasmhughes** [comment]
re-run all

**Comment by MKLeb** [comment]
@nicholasmhughes Sorry, CI is currently broken but there is a PR up to fix it.  I'll rerun after that is merged and then likely approve this.

**Comment by nicholasmhughes** [comment]
re-run pr-ubuntu-2004-arm64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-centos-8-x86_64-py3-pytest

**Comment by MKLeb** [comment]
test failure looks unrelated, will review then

**Review by MKLeb — DISMISSED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-centos-7-x86_64-py3-m2crypto-pytest

**Comment by nicholasmhughes** [comment]
@MKLeb is everything good with this PR?

**Comment by MKLeb** [comment]
Looks good on my end.  I've sent it around internally to have others take a look as well, but if there are no concerns then it should be good to go.

File: tests/pytests/functional/fileserver/test_roots.py [hunk 2]
```diff
@@ -9,8 +14,12 @@
 
 @pytest.fixture(scope="module")
 def configure_loader_modules(base_env_state_tree_root_dir):
+    cachedir = os.path.join(RUNTIME_VARS.TMP, "__salt_test_fileserver_roots_cache_dir")
+    if not os.path.isdir(cachedir):
+        os.makedirs(cachedir)
     opts = salt.config.DEFAULT_MINION_OPTS.copy()
-    print(base_env_state_tree_root_dir)
+    print(base_env_state_tree_root_dir, end=" ")
```
Comment (MKLeb) [inline tests/pytests/functional/fileserver/test_roots.py:21]: Is there any purpose to this `print`?  I wonder if it was leftover from versions prior that just never got caught.

File: tests/pytests/functional/fileserver/test_roots.py [hunk 3]
```diff
@@ -9,8 +14,12 @@
 
 @pytest.fixture(scope="module")
 def configure_loader_modules(base_env_state_tree_root_dir):
+    cachedir = os.path.join(RUNTIME_VARS.TMP, "__salt_test_fileserver_roots_cache_dir")
+    if not os.path.isdir(cachedir):
+        os.makedirs(cachedir)
```
Comment (MKLeb) [inline tests/pytests/functional/fileserver/test_roots.py:19]:
```suggestion
def configure_loader_modules(base_env_state_tree_root_dir, tmp_path):
    cachedir = tmp_path / "__salt_test_fileserver_roots_cache_dir"
    cachedir.mkdir(parents=True, exist_ok=True)
```

File: tests/pytests/functional/fileserver/test_roots.py [hunk 4]
```diff
@@ -9,8 +14,12 @@
...
     opts = salt.config.DEFAULT_MINION_OPTS.copy()
-    print(base_env_state_tree_root_dir)
+    print(base_env_state_tree_root_dir, end=" ")
+    opts["cachedir"] = cachedir
```
Comment (MKLeb) [inline tests/pytests/functional/fileserver/test_roots.py:22]: To get along with the above suggested changes...
```suggestion
    opts["cachedir"] = str(cachedir)
```

File: tests/pytests/functional/fileserver/test_roots.py (hunk 3 above)
Comment (s0undt3ch) [reply tests/pytests/functional/fileserver/test_roots.py:19]: :point_up:

File: tests/pytests/functional/fileserver/test_roots.py (hunk 2 above)
Comment (s0undt3ch) [reply tests/pytests/functional/fileserver/test_roots.py:21]: Yeah, let's nuke it

**Review by s0undt3ch — CHANGES_REQUESTED** [review]

**Review by MKLeb — APPROVED** [review]

**Review by twangboy — APPROVED** [review]

**Review by s0undt3ch — APPROVED** [review]

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by nicholasmhughes** [comment]
re-run pr-arch-lts-x86_64-py3-pytest

**Comment by morgana2313** [comment]
I'm happy to see my ideas from #55747 were implemented and merged! 👍 

I think this new `__env__` behaviour should also land in the pillar.stack:

```
ext_pillar:
  - stack:
      pillarenv:
        __env__: /path/__env__/pillar/stack.cfg
        __env__: /path2/__env__/pillar/stack.cfg
```

My PR #55812 shows how this might be accomplished:

```diff
- stack_config_files += cfgs
+ stack_config_files += [cfg.replace("__env__", pillarenv) for cfg in cfgs]
```

@nicholasmhughes would you be willing to cooperate on this?

**Comment by nicholasmhughes** [comment]
@morgana2313 , happy to help in any way I can! Feel free to reach out on [Slack](https://saltproject.io/contact/) if it's easier to coordinate that way.

**Pull request merged**